l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,638 @@
+            require(withdrawAmount == _depositAmount, "5T"); // The token has non-standard transfer logic
+        }
+        require(amount != 0, "6T"); // empty deposit amount
+
+        bytes32 txDataHash = keccak256(abi.encode(_prevMsgSender, _l1Token, amount));
+        if (!hyperbridgingEnabled[_chainId]) {
+            chainBalance[_chainId][_l1Token] += amount;
+        }
+
+        {
+            // Request the finalization of the deposit on the L2 side
+            bytes memory l2TxCalldata = _getDepositL2Calldata(_prevMsgSender, _l2Receiver, _l1Token, amount);
+
+            request = L2TransactionRequestTwoBridgesInner({
+                magicValue: TWO_BRIDGES_MAGIC_VALUE,
+                l2Contract: l2BridgeAddress[_chainId],
+                l2Calldata: l2TxCalldata,
+                factoryDeps: new bytes[](0),
+                txDataHash: txDataHash
+            });
+        }
+        emit BridgehubDepositInitiatedSharedBridge(_chainId, txDataHash, _prevMsgSender, _l2Receiver, _l1Token, amount);
+        // kl todo. We are breaking the previous events here, as we don't have the txHash, so we can't emit a DepositInitiated event
+    }
+
+    /// @notice Confirms the acceptance of a transaction by the Mailbox, as part of the L2 transaction process within Bridgehub.
+    /// This function is utilized by `requestL2TransactionTwoBridges` to validate the execution of a transaction.
+    function bridgehubConfirmL2Transaction(
+        uint256 _chainId,
+        bytes32 _txDataHash,
+        bytes32 _txHash
+    ) external override onlyBridgehub {
+        require(depositHappened[_chainId][_txHash] == 0x00, "ShB tx hap");
+        depositHappened[_chainId][_txHash] = _txDataHash;
+        emit BridgehubDepositFinalized(_chainId, _txDataHash, _txHash);
+    }
+
+    /// @dev Generate a calldata for calling the deposit finalization on the L2 bridge contract
+    function _getDepositL2Calldata(
+        address _l1Sender,
+        address _l2Receiver,
+        address _l1Token,
+        uint256 _amount
+    ) internal view returns (bytes memory) {
+        bytes memory gettersData = _getERC20Getters(_l1Token);
+        return abi.encodeCall(IL2Bridge.finalizeDeposit, (_l1Sender, _l2Receiver, _l1Token, _amount, gettersData));
+    }
+
+    /// @dev Receives and parses (name, symbol, decimals) from the token contract
+    function _getERC20Getters(address _token) internal view returns (bytes memory) {
+        if (_token == ETH_TOKEN_ADDRESS) {
+            return abi.encode("Ether", "ETH", uint8(18)); // when depositing eth to a non-eth based chain it is an ERC20
+        }
+
+        (, bytes memory data1) = _token.staticcall(abi.encodeCall(IERC20Metadata.name, ()));
+        (, bytes memory data2) = _token.staticcall(abi.encodeCall(IERC20Metadata.symbol, ()));
+        (, bytes memory data3) = _token.staticcall(abi.encodeCall(IERC20Metadata.decimals, ()));
+        return abi.encode(data1, data2, data3);
```

Comment:
EDGE CASE: The IERC20Metadata.name or .symbol function from OZ libraries expect the return values to be of type string. See [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/IERC20Metadata.sol), but it is not always the case that tokens return string values, see [edge case here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/IERC20Metadata.sol)

Reply:
Yes, the issue is, there are tokens that **do implement** `name` and `symbol` functions, but instead of returning a `string` as expected, they return a `bytes32` value and that is something that the OZ functions (namely `IERC20Metadata.name` and `IERC20Metadata.symbol` are not equipped to handle).

Am I making sense here? 🥹

Reply:
@saxenism I know about Maker token, and don't see any issues here. They indeed return `bytes32` instead of `string` and this is how we handle it:
1) `(, bytes memory data1) = _token.staticcall(abi.encodeCall(IERC20Metadata.name, ()));` - Bridge pull the data on L1.
2) L2 bridge counterpart **try** to decode the data, decoding failed and so standard value would be returned on `name` instead of `bytes32` value that is stored on L1 token.

All in all, token that support `ERC-20` interface in wrong way will have standard `name` value instead of custom one - lgtm